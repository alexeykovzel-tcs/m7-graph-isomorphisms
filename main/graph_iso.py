from color_ref_fast import GraphFastRef
from color_ref import GraphColors
from graph import Graph
import sys


detect_forests = True
detect_false_twins = True
use_fast_refinement = True


class IsoGroup:
    def __init__(self, graphs: set[Graph], count=0):
        self.graphs = graphs
        self.count = count


class GraphIso:
    def __init__(self, graphs: set[Graph]):
        self.graphs = graphs

        # detect graph forests
        if detect_forests:
            self.forests = set(g for g in graphs if g.is_forest())

        # detect false twins
        if detect_false_twins:
            self.false_twins = self.__false_twins()

    def group(self, with_count=False):
        grouped_graphs = set()
        groups = []

        # find all isomorphic groups
        colorRef = GraphFastRef(self.graphs) \
            if use_fast_refinement else GraphColors(self.graphs)
        colors = colorRef.reset().refine()
        for color_group in colors.group():
            graphs = color_group.graphs
            for i in range(len(graphs) - 1):
                if graphs[i] in grouped_graphs:
                    continue

                # create a new isomorphic group
                group = IsoGroup({graphs[i]})
                groups.append(group)
                for j in range(i + 1, len(graphs)):
                    pair = [graphs[i], graphs[j]]
                    main_branch = colors.copy(pair)

                    # check if graphs are isomorphic
                    isomorphic = False
                    if with_count and group.count == 0:
                        count = self.__count(pair, main_branch)
                        if count > 0:
                            group.count = count
                            isomorphic = True
                    else:
                        count = self.__count(pair, main_branch, only_one=True)
                        isomorphic = count == 1

                    if isomorphic:
                        group.graphs.add(graphs[j])
                        grouped_graphs.add(graphs[j])
        
        # handle signle graph groups
        for group in groups:
            if len(group.graphs) == 1 and group.count == 0:
                graph = list(group.graphs)[0]
                pair = [graph, graph.deep_copy()]
                new_count = GraphIso(pair).group(with_count=True)[0].count
                group.count = new_count

        return groups

    def __count(self, graphs, colors, only_one=False):
        color_groups = colors.group()
        total_count = 0

        # if coloring is unbalanced, there are no isomorphisms
        if len(color_groups) == 0:
            return 0

        # if coloring defines a bijection, there is only 1 automorphism
        if color_groups[0].discrete:
            return 1

        # branching rules: choosing the color class
        color_set = color_groups[0].colors
        branch_color = min(color_set, key=lambda c: color_set[c] if color_set[c] > 1 else sys.maxsize)

        # get x and y vertices
        y_all = colors.vertices_of(graphs[1], branch_color)
        x = colors.first_vertex_of(graphs[0], branch_color)

        # count isomorphisms for each possible branch
        y_precounted = {}
        for y in y_all:
            if y in y_precounted:
                total_count += y_precounted[y]
                continue

            # calculate isomorphisms for this branch
            branch = colors.copy(graphs)
            branch.assign(x, branch.next)
            branch.assign(y, branch.next)

            if use_fast_refinement:
                next_color = branch.colors[branch.next]
                branch.queue.append(next_color)

            branch.next += 1
            branch.refine()

            # calculate isomorphisms for this branch
            count = self.__count(graphs, branch, only_one)
            if only_one and count == 1:
                return 1

            # copy results for forests
            if detect_forests and graphs[0] in self.forests:
                y_precounted = {v: count for v in y_all}

            # copy results for false twins
            if detect_false_twins:
                for vertex in self.false_twins[y]:
                    y_precounted[vertex] = count

            total_count += count

        return total_count

    def __false_twins(self):
        false_twins = {}
        for graph in self.graphs:
            for v1 in graph.vertices:
                false_twins[v1] = set()
                for v2 in graph.vertices:

                    n1 = set(v1._incidence.keys())
                    n2 = set(v2._incidence.keys())

                    if v1 != v2 and n1 == n2 and not v1.is_adjacent(v2):
                        false_twins[v1].add(v2)

        return false_twins
