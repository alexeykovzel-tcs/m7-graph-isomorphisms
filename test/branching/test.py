import sys, os

# appends the system path to the graph files
sys.path.append(os.path.abspath('main'))
sys.path.append(os.path.abspath('test'))

from test_utils import format_groups, test
from graph_iso import GraphIso
from graph import Graph


expected_path = 'test/branching/test_expected.txt'
output_path = 'test/branching/test_actual.txt'
samples_path = 'test/branching/samples/'
graph_files = [
    # 'torus24.grl',
    # 'torus72.grl',
    # 'torus144.grl',
    # 'products72.grl',
    # 'products216.grl',
    # 'cographs1.grl',
    # 'trees11.grl',
    # 'trees36.grl',
    # 'trees90.grl',
    'modulesC.grl',
    # 'modulesD.grl',
    # 'cubes3.grl',
    # 'cubes5.grl',
    # 'cubes6.grl',
    # 'cubes7.grl',
    # 'cubes9.grl',
    # 'bigtrees1.grl',
    # 'bigtrees2.grl',
    # 'bigtrees3.grl',
    # 'wheeljoin14.grl',
    # 'wheeljoin33.grl',
    # 'wheelstar12.grl',
    # 'wheelstar15.grl',
]


def iso_groups_out(f, graphs, groups):
    format_groups(graphs, groups)
    output = f + ':\n'
    for group in groups:
        output += '{} {}\n'.format(str(group.graphs), str(group.count))
    return output

def iso_groups(graphs: set[Graph]):
    return GraphIso(graphs).group(with_count=True)


if __name__ == "__main__":
    f = open(expected_path, "r")
    expected = f.read().split('\n\n')[1:]

    out = 'Isomorphic pairs:\n\n'
    out += test(samples_path, [(
        graph_files, 
        iso_groups,
        iso_groups_out
    )], expected)

    with open(output_path, 'w') as f:
        f.write(out)
