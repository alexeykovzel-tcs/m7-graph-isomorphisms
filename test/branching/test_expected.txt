Isomorphic pairs:

torus24.grl:
[0, 3] 96
[1, 2] 96

torus72.grl:
[0, 2] 288
[1, 5] 288
[3, 6] 288
[4, 7] 288

torus144.grl:
[0, 6] 576
[1, 7] 576
[2, 4] 576
[3, 10] 576
[5, 9] 1152
[8, 11] 576

products72.grl:
[0, 6] 288
[1, 5] 576
[2, 3] 576
[4, 7] 864

products216.grl:
[0, 6] 1728
[1, 7] 1728
[2, 9] 1728
[3, 8] 10368
[4, 5] 1728

cographs1.grl:
[0, 3] 5971968
[1, 2] 995328

trees11.grl:
[0, 3] 6
[1, 4] 1
[2, 5] 2

trees36.grl:
[0, 7] 2
[1, 4] 6
[2, 6] 2
[3, 5] 6

trees90.grl:
[0, 3] 6912
[1, 2] 20736

modulesC.grl:
[0, 7] 17915904
[1, 5] 17915904
[2, 4] 2488320
[3, 6] 2985984

modulesD.grl:
[0, 2] 24
[1, 3] 1
[4, 5] 24

cubes3.grl:
[0, 2] 48
[1, 3] 16

cubes5.grl:
[0, 1] 3840
[2, 3] 24

cubes6.grl:
[0, 1] 96
[2, 3] 46080

cubes7.grl:
[0, 3] 645120
[1, 2] 480

cubes9.grl:
[0, 1] 185794560
[2, 3] 20160

bigtrees1.grl:
[0, 2] 442368
[1, 3] 5308416

bigtrees2.grl:
[0, 3] 80244904034304
[1, 2] 160489808068608

bigtrees3.grl:
[0, 2] 2772351862699137701073289910157312
[1, 3] 462058643783189616845548318359552

wheeljoin14.grl:
[0, 1] 1600
[2, 3] 672
[4, 7] 1536
[5, 6] 720

wheeljoin33.grl:
[0, 4] 8257536
[1, 2] 7962624
[3, 5] 50577408
[6, 7] 1290240

whellstar12.grl:
[0, 3] 1935360
[1, 2] 6718464

wheelstar15.grl:
[0, 7] 1703116800
[1, 4] 3009871872
[2, 3] 10642046976
[5, 6] 2890137600